File: packages/capacitor/android/src/main/java/com/bendyline/gezel/capacitor/GezelRuntimePlugin.java
package com.bendyline.gezel.capacitor;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import androidx.activity.result.ActivityResult;
import com.getcapacitor.*;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.bendyline.gezel.runtime.GezelNativeRuntime;
import com.bendyline.gezel.runtime.NativeCall;
import org.json.JSONObject;

@CapacitorPlugin(name = "GezelRuntime")
public final class GezelRuntimePlugin extends Plugin {
    private GezelNativeRuntime runtime;
    private AutoCloseable listener;
    private boolean pickingModel;
    @Override public void load() {
        runtime = GezelNativeRuntime.shared(getContext());
        listener = runtime.listen((name, data) -> notifyListeners(name, toJS(data)));
    }
    public static JSObject toJS(JSONObject value) {
        JSObject copy = new JSObject();
        for (var keys = value.keys(); keys.hasNext();) { String key = keys.next(); copy.put(key, value.opt(key)); }
        return copy;
    }
    public static NativeCall adapt(PluginCall call) {
        return new NativeCall(call.getData(), new NativeCall.Reply() {
            public void resolve(JSONObject value) { call.resolve(toJS(value)); }
            public void reject(String message, String code) { call.reject(message, code); }
        });
    }
    @PluginMethod public void providers(PluginCall call) { runtime.providers(adapt(call)); }
    @PluginMethod public void prepareProvider(PluginCall call) { runtime.prepareProvider(adapt(call)); }
    @PluginMethod public void cancelProviderPreparation(PluginCall call) { runtime.cancelProviderPreparation(adapt(call)); }
    @PluginMethod public void generate(PluginCall call) { runtime.generate(adapt(call)); }
    @PluginMethod public void cancel(PluginCall call) { runtime.cancel(adapt(call)); }
    @PluginMethod public void releaseModel(PluginCall call) { runtime.releaseModel(adapt(call)); }
    @PluginMethod public void listModels(PluginCall call) { runtime.listModels(adapt(call)); }
    @PluginMethod public void selectModel(PluginCall call) { runtime.selectModel(adapt(call)); }
    @PluginMethod public void removeModel(PluginCall call) { runtime.removeModel(adapt(call)); }
    @PluginMethod public void resolveModelSource(PluginCall call) { runtime.resolveModelSource(adapt(call)); }
    @PluginMethod public void cancelModelSourceResolution(PluginCall call) { runtime.cancelModelSourceResolution(adapt(call)); }
    @PluginMethod public void listModelDownloads(PluginCall call) { runtime.listModelDownloads(adapt(call)); }
    @PluginMethod public void startModelDownload(PluginCall call) { runtime.startModelDownload(adapt(call)); }
    @PluginMethod public void resumeModelDownload(PluginCall call) { runtime.resumeModelDownload(adapt(call)); }
    @PluginMethod public void cancelModelDownload(PluginCall call) { runtime.cancelModelDownload(adapt(call)); }
    @PluginMethod public void removeModelDownload(PluginCall call) { runtime.removeModelDownload(adapt(call)); }
    @PluginMethod public void importModel(PluginCall call) {
        if (!runtime.reserveModelMutation()) { call.reject("Finish the current operation before importing", "BUSY"); return; }
        pickingModel = true;
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        intent.setType("application/octet-stream");
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        try { startActivityForResult(call, intent, "modelPicked"); }
        catch (Exception error) { pickingModel = false; runtime.releaseModelMutation(); call.reject("Document picker unavailable", error); }
    }
    @ActivityCallback private void modelPicked(PluginCall call, ActivityResult result) {
        if (!pickingModel) { if (call != null) call.reject("Document picker closed", "CANCELLED"); return; }
        pickingModel = false;
        runtime.releaseModelMutation();
        if (call == null) return;
        Uri uri = result.getData() == null ? null : result.getData().getData();
        if (result.getResultCode() != Activity.RESULT_OK || uri == null) { call.resolve(new JSObject().put("model", JSONObject.NULL)); return; }
        runtime.importModel(adapt(call), uri);
    }
    @Override protected void handleOnStop() { runtime.onBackground(); }
    @Override protected void handleOnResume() { runtime.onForeground(); }
    @Override protected void handleOnDestroy() {
        if (pickingModel) { pickingModel = false; runtime.releaseModelMutation(); }
        try { if (listener != null) listener.close(); } catch (Exception ignored) {}
    }
}

File: packages/capacitor/ios/Sources/GezelCapacitor/GezelRuntimePlugin.swift
import Capacitor
import Foundation
import UIKit
import UniformTypeIdentifiers
import WebKit
#if canImport(GezelRuntime)
import GezelRuntime
#endif

@objc(GezelRuntimePlugin)
public final class GezelRuntimePlugin: CAPPlugin, CAPBridgedPlugin, UIDocumentPickerDelegate {
    public let identifier = "GezelRuntimePlugin"
    public let jsName = "GezelRuntime"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "listModels", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resolveModelSource", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelModelSourceResolution", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listModelDownloads", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startModelDownload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeModelDownload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelModelDownload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeModelDownload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "providers", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prepareProvider", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelProviderPreparation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "selectModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "importModel", returnType: CAPPluginReturnPromise)
    ]
    private var runtime: GezelNativeRuntime?
    private var listener: UUID?
    private var pickerCall: CAPPluginCall?
    public override func load() {
        runtime = try? GezelNativeRuntime.shared()
        listener = runtime?.listen { [weak self] event, data in self?.notifyListeners(event, data: data) }
    }
    deinit {
        if let listener { runtime?.removeListener(listener) }
        if pickerCall != nil { runtime?.releaseModelMutation() }
    }
    public static func adapt(_ call: CAPPluginCall) -> NativeCall {
        NativeCall(call.options as? [String: Any] ?? [:], resolve: { call.resolve($0) }, reject: { call.reject($0, $1) })
    }
    private func withRuntime(_ call: CAPPluginCall, _ action: (GezelNativeRuntime) -> Void) {
        guard let runtime else { call.reject("Native runtime initialization failed", "UNAVAILABLE"); return }
        action(runtime)
    }
    @objc public func listModels(_ call: CAPPluginCall) { withRuntime(call) { $0.listModels(Self.adapt(call)) } }
    @objc public func resolveModelSource(_ call: CAPPluginCall) { withRuntime(call) { $0.resolveModelSource(Self.adapt(call)) } }
    @objc public func cancelModelSourceResolution(_ call: CAPPluginCall) { withRuntime(call) { $0.cancelModelSourceResolution(Self.adapt(call)) } }
    @objc public func listModelDownloads(_ call: CAPPluginCall) { withRuntime(call) { $0.listModelDownloads(Self.adapt(call)) } }
    @objc public func startModelDownload(_ call: CAPPluginCall) { withRuntime(call) { $0.startModelDownload(Self.adapt(call)) } }
    @objc public func resumeModelDownload(_ call: CAPPluginCall) { withRuntime(call) { $0.resumeModelDownload(Self.adapt(call)) } }
    @objc public func cancelModelDownload(_ call: CAPPluginCall) { withRuntime(call) { $0.cancelModelDownload(Self.adapt(call)) } }
    @objc public func removeModelDownload(_ call: CAPPluginCall) { withRuntime(call) { $0.removeModelDownload(Self.adapt(call)) } }
    @objc public func providers(_ call: CAPPluginCall) { withRuntime(call) { $0.providers(Self.adapt(call)) } }
    @objc public func prepareProvider(_ call: CAPPluginCall) { withRuntime(call) { $0.prepareProvider(Self.adapt(call)) } }
    @objc public func cancelProviderPreparation(_ call: CAPPluginCall) { withRuntime(call) { $0.cancelProviderPreparation(Self.adapt(call)) } }
    @objc public func selectModel(_ call: CAPPluginCall) { withRuntime(call) { $0.selectModel(Self.adapt(call)) } }
    @objc public func removeModel(_ call: CAPPluginCall) { withRuntime(call) { $0.removeModel(Self.adapt(call)) } }
    @objc public func generate(_ call: CAPPluginCall) { withRuntime(call) { $0.generate(Self.adapt(call)) } }
    @objc public func cancel(_ call: CAPPluginCall) { withRuntime(call) { $0.cancel(Self.adapt(call)) } }
    @objc public func releaseModel(_ call: CAPPluginCall) { withRuntime(call) { $0.releaseModel(Self.adapt(call)) } }
    @objc public func importModel(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let runtime = self.runtime, runtime.reserveModelMutation() else { call.reject("Finish the current operation before importing", "BUSY"); return }
            guard let controller = self.bridge?.viewController, controller.presentedViewController == nil else { runtime.releaseModelMutation(); call.reject("Document picker unavailable"); return }
            self.pickerCall = call
            let picker = UIDocumentPickerViewController(forOpeningContentTypes: [.data], asCopy: false)
            picker.allowsMultipleSelection = false; picker.delegate = self
            controller.present(picker, animated: true)
        }
    }
    public func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        guard let call = pickerCall else { return }; pickerCall = nil
        runtime?.releaseModelMutation(); call.resolve(["model": NSNull()])
    }
    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let call = pickerCall else { return }; pickerCall = nil
        runtime?.releaseModelMutation()
        guard let url = urls.first else { call.resolve(["model": NSNull()]); return }
        withRuntime(call) { $0.importModel(Self.adapt(call), from: url) }
    }
}
